### PR TITLE
Update the BNN tutorial #3

### PR DESCRIPTION
**Pull request by xukai92**

### CODE REVIEW

File: tutorials/03-bayesian-neural-network/03_bayesian-neural-network.jmd [hunk 1]
```diff
@@ -62,54 +62,42 @@ plot_data()
 
 ## Building a Neural Network
 
-The next step is to define a [feedforward neural network](https://en.wikipedia.org/wiki/Feedforward_neural_network) where we express our parameters as distribtuions, and not single points as with traditional neural networks. The two functions below, `unpack` and `nn_forward` are helper functions we need when we specify our model in Turing.
-
-`unpack` takes a vector of parameters and partitions them between weights and biases. `nn_forward` constructs a neural network with the variables generated in `unpack` and returns a prediction based on the weights provided.
-
-The `unpack` and `nn_forward` functions are explicity designed to create a neural network with two hidden layers and one output layer, as shown below.
+The next step is to define a [feedforward neural network](https://en.wikipedia.org/wiki/Feedforward_neural_network) where we express our parameters as distribtuions, and not single points as with traditional neural networks.
+For this we will use `Dense` to define liner layers and compose them via `Chain`, both are neural network primitives from from Flux.
```
Comment (yebai) [inline tutorials/03-bayesian-neural-network/03_bayesian-neural-network.jmd:66]:
```suggestion
For this we will use `Dense` to define liner layers and compose them via `Chain`, both are neural network primitives from Flux.
```

File: tutorials/03-bayesian-neural-network/03_bayesian-neural-network.jmd [hunk 2]
```diff
@@ -62,54 +62,42 @@ plot_data()
 
 ## Building a Neural Network
 
-The next step is to define a [feedforward neural network](https://en.wikipedia.org/wiki/Feedforward_neural_network) where we express our parameters as distribtuions, and not single points as with traditional neural networks. The two functions below, `unpack` and `nn_forward` are helper functions we need when we specify our model in Turing.
-
-`unpack` takes a vector of parameters and partitions them between weights and biases. `nn_forward` constructs a neural network with the variables generated in `unpack` and returns a prediction based on the weights provided.
-
-The `unpack` and `nn_forward` functions are explicity designed to create a neural network with two hidden layers and one output layer, as shown below.
+The next step is to define a [feedforward neural network](https://en.wikipedia.org/wiki/Feedforward_neural_network) where we express our parameters as distribtuions, and not single points as with traditional neural networks.
```
Comment (yebai) [inline tutorials/03-bayesian-neural-network/03_bayesian-neural-network.jmd:65]:
```suggestion
The next step is to define a [feedforward neural network](https://en.wikipedia.org/wiki/Feedforward_neural_network) where we express our parameters as distributions, and not single points as with traditional neural networks.
```

File: tutorials/03-bayesian-neural-network/03_bayesian-neural-network.jmd [hunk 3]
```diff
@@ -62,54 +62,42 @@ plot_data()
...
-function nn_forward(xs, nn_params::AbstractVector)
-    W₁, b₁, W₂, b₂, Wₒ, bₒ = unpack(nn_params)
-    nn = Chain(Dense(W₁, b₁, tanh), Dense(W₂, b₂, tanh), Dense(Wₒ, bₒ, σ))
-    return nn(xs)
-end;
+length(parameters_initial) # number of paraemters in NN
 ```
 
-The probabilistic model specification below creates a `params` variable, which has 20 normally distributed variables. Each entry in the `params` vector represents weights and biases of our neural net.
+The probabilistic model specification below creates a `parameters` variable, which has IID normal variables. The `parameters` vector represents all paraemters of our neural net (weights and biases).
```
Comment (yebai) [inline tutorials/03-bayesian-neural-network/03_bayesian-neural-network.jmd:85]:
```suggestion
The probabilistic model specification below creates a `parameters` variable, which has IID normal variables. The `parameters` vector represents all parameters of our neural net (weights and biases).
```

File: tutorials/03-bayesian-neural-network/03_bayesian-neural-network.jmd [hunk 4]
```diff
@@ -62,54 +62,42 @@ plot_data()
...
 
 ```julia
 # Create a regularization term and a Gaussain prior variance term.
```
Comment (yebai) [inline tutorials/03-bayesian-neural-network/03_bayesian-neural-network.jmd:88]:
```suggestion
# Create a regularization term and a Gaussian prior variance term.
```

File: tutorials/03-bayesian-neural-network/03_bayesian-neural-network.jmd [hunk 5]
```diff
@@ -123,21 +111,27 @@ Inference can now be performed by calling `sample`. We use the `HMC` sampler her
...
 N = 5000
-ch = sample(bayes_nn(hcat(xs...), ts), HMC(0.05, 4), N);
+ch = sample(
+    bayes_nn(hcat(xs...), ts, length(parameters_initial), reconstruct), HMC(0.05, 4), N
+);
 ```
 
-Now we extract the weights and biases from the sampled chain. We'll use these primarily in determining how good a classifier our model is.
+Now we extract the parameter samples from the sampled chain as `theta` (this is of size `5000 x 20` where `5000` is the number of iterations and `20` is the number of parameters).
+We'll use these primarily in determining how good a classifier our model is.
```
Comment (yebai) [inline tutorials/03-bayesian-neural-network/03_bayesian-neural-network.jmd:120]:
```suggestion
We'll use these primarily to determine how good our model's classifier is.
```

File: tutorials/03-bayesian-neural-network/03_bayesian-neural-network.jmd [hunk 6]
```diff
@@ -172,17 +166,17 @@ function nn_predict(x, theta, num)
 end;
 ```
 
-Next, we use the `nn_predict` function to predict the value at a sample of points where the `x` and `y` coordinates range between -6 and 6. As we can see below, we still have a satisfactory fit to our data.
+Next, we use the `nn_predict` function to predict the value at a sample of points where the `x1` and `x2` coordinates range between -6 and 6. As we can see below, we still have a satisfactory fit to our data, and more importantly, we can also see where the neural network is uncertain about its predictions much easier---those region between cluster boundaries.
```
Comment (yebai) [inline tutorials/03-bayesian-neural-network/03_bayesian-neural-network.jmd:169]:
```suggestion
Next, we use the `nn_predict` function to predict the value at a sample of points where the `x1` and `x2` coordinates range between -6 and 6. As we can see below, we still have a satisfactory fit to our data, and more importantly, we can also see where the neural network is uncertain about its predictions much easier--- those regions between cluster boundaries.
```

File: tutorials/03-bayesian-neural-network/03_bayesian-neural-network.jmd [hunk 7]
```diff
@@ -172,17 +166,17 @@ function nn_predict(x, theta, num)
...
-y_range = collect(range(-6; stop=6, length=25))
-Z = [nn_predict([x, y], theta, n_end)[1] for x in x_range, y in y_range]
-contour!(x_range, y_range, Z)
+x1_range = collect(range(-6; stop=6, length=25))
+x2_range = collect(range(-6; stop=6, length=25))
+Z = [nn_predict([x1, x2], theta, n_end)[1] for x1 in x1_range, x2 in x2_range]
+contour!(x1_range, x2_range, Z)
 ```
 
 If you are interested in how the predictive power of our Bayesian neural network evolved between samples, the following graph displays an animation of the contour plot generated from the network weights in samples 1 to 1,000.
```
Comment (yebai) [inline tutorials/03-bayesian-neural-network/03_bayesian-neural-network.jmd:182]:
```suggestion
Suppose you are interested in how the predictive power of our Bayesian neural network evolved between samples. In that case, the following graph displays an animation of the contour plot generated from the network weights in samples 1 to 1,000.
```

**Review by yebai — CHANGES_REQUESTED** [review]
Many thanks, @xukai92 - it looks good. I left some minor comments below.

**Comment by xukai92** [comment]
Thanks for the review---I applied the changes. Shall we merge this PR?

**Comment by yebai** [comment]
Sure, sounds good to merge it.

**Pull request merged**